File: packages/are-mcp-gateway/src/middleware.ts
import { asMcpToolResult, evaluateToolCall } from "./foundation.js";
import type { AreGatewayConfig, McpToolResult, ToolCall } from "./types.js";

type AnyFunction = (...args: unknown[]) => unknown | Promise<unknown>;

export function governMcpTool<T extends AnyFunction>(tool: T, config: AreGatewayConfig): T {
  const governed = async (...args: Parameters<T>): Promise<Awaited<ReturnType<T>> | McpToolResult> => {
    const call: ToolCall = {
      name: tool.name || "mcp.tool",
      args: args.length <= 1 ? args[0] : args
    };
    const decision = await evaluateToolCall(call, config);
    if (decision.enforcedEffect !== "ALLOW") {
      return asMcpToolResult(decision);
    }
    return (await tool(...args)) as Awaited<ReturnType<T>>;
  };
  Object.defineProperty(governed, "name", { value: `governed_${tool.name || "tool"}` });
  return governed as T;
}

export async function runGovernedTool<T>(
  call: ToolCall,
  config: AreGatewayConfig,
  invoke: () => Promise<T> | T
): Promise<T | McpToolResult> {
  const decision = await evaluateToolCall(call, config);
  if (decision.enforcedEffect !== "ALLOW") {
    return asMcpToolResult(decision);
  }
  return await invoke();
}

File: python/src/are_mcp_gateway/middleware.py
from __future__ import annotations

from functools import wraps
from typing import Any, Callable, Mapping

from .client import AreDecision, AreGatewayConfig, evaluate_tool_call


def govern_tool(
    tool: Callable[..., Any],
    *,
    foundation_url: str,
    token: str,
    agent_id: str,
    map_tool_call: Callable[[Mapping[str, Any]], Mapping[str, Any]],
    passport_id: str | None = None,
    mode: str = "enforce",
    timeout_seconds: float = 2.5,
) -> Callable[..., Any]:
    config = AreGatewayConfig(
        foundation_url=foundation_url,
        token=token,
        agent_id=agent_id,
        passport_id=passport_id,
        mode="observe" if mode == "observe" else "enforce",
        timeout_seconds=timeout_seconds,
        map_tool_call=map_tool_call,
    )

    @wraps(tool)
    def wrapped(*args: Any, **kwargs: Any) -> Any:
        call = {"name": getattr(tool, "__name__", "mcp.tool"), "args": {"args": args, "kwargs": kwargs}}
        decision = evaluate_tool_call(call, config)
        if decision.enforced_effect != "ALLOW":
            return as_tool_error(decision)
        return tool(*args, **kwargs)

    return wrapped


def run_governed_tool(call: Mapping[str, Any], config: AreGatewayConfig, invoke: Callable[[], Any]) -> Any:
    decision = evaluate_tool_call(call, config)
    if decision.enforced_effect != "ALLOW":
        return as_tool_error(decision)
    return invoke()


def as_tool_error(decision: AreDecision) -> dict[str, Any]:
    label = "requires approval" if decision.effect == "ESCALATE" else "denied"
    return {
        "is_error": True,
        "content": [{"type": "text", "text": f"ARE {label} this tool call: {decision.reason}"}],
        "are_decision": {
            "effect": decision.effect,
            "enforced_effect": decision.enforced_effect,
            "reason": decision.reason,
            "request_id": decision.request_id,
            "executed": False,
        },
    }
